File: laravel-backend/app/Models/Transaction.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

/**
 * Transaction Model
 * Matches Next.js and Flutter transaction structure
 */
class Transaction extends Model
{
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'date',
        'type',
        'category',
        'description',
        'amount',
        'payment_method',
        'transaction_number',
        'receipt_number',
        'tin_number',
        'vat',
        'supplier_name',
        'supplier_address',
    ];

    protected $casts = [
        'date' => 'date',
        'amount' => 'decimal:2',
        'vat' => 'integer',
    ];

    // Transaction types
    const TYPE_REVENUE = 'revenue';
    const TYPE_TRANSACTION = 'transaction';

    // Revenue categories (matching Next.js REVENUE_CATEGORIES)
    const REVENUE_CATEGORIES = [
        'Cash',
        'GCash',
        'Grab',
        'PayMaya',
        'Others',
    ];

    // Transaction categories (matching Next.js TRANSACTION_CATEGORIES)
    const TRANSACTION_CATEGORIES = [
        'Supplies',
        'Pastries',
        'Rent',
        'Utilities',
        'Manpower',
        'Marketing',
        'Others',
    ];

    // Payment methods (matching Next.js PAYMENT_METHODS)
    const PAYMENT_METHODS = [
        'Cash',
        'Check',
        'Bank Transfer',
        'Credit Card',
        'GCash',
        'PayMaya',
        'Others',
    ];

    /**
     * Scope for revenue transactions
     */
    public function scopeRevenue($query)
    {
        return $query->where('type', self::TYPE_REVENUE);
    }

    /**
     * Scope for transaction (expense) transactions
     */
    public function scopeTransactionType($query)
    {
        return $query->where('type', self::TYPE_TRANSACTION);
    }

    /**
     * Scope for date range
     */
    public function scopeDateRange($query, $startDate, $endDate)
    {
        return $query->whereBetween('date', [$startDate, $endDate]);
    }

    /**
     * Scope for specific category
     */
    public function scopeCategory($query, $category)
    {
        return $query->where('category', $category);
    }

    /**
     * Calculate VAT amount
     */
    public function getVatAmountAttribute()
    {
        return $this->amount * ($this->vat / 100);
    }

    /**
     * Get amount including VAT
     */
    public function getAmountWithVatAttribute()
    {
        return $this->amount + $this->vat_amount;
    }
}

File: lib/models/transaction.dart
import 'package:json_annotation/json_annotation.dart';

part 'transaction.g.dart';

/// Transaction model matching Next.js transaction structure
/// Maps from: { id, date, type, category, description, amount, paymentMethod, transactionNumber, receiptNumber, tinNumber, vat, supplierName, supplierAddress }
@JsonSerializable()
class Transaction {
  final int id;
  final String date; // ISO format: "2025-11-09"
  final TransactionType type;
  final String category;
  final String description;
  final double amount;
  @JsonKey(name: 'payment_method')
  final String paymentMethod;
  @JsonKey(name: 'transaction_number')
  final String transactionNumber;
  @JsonKey(name: 'receipt_number')
  final String receiptNumber;
  @JsonKey(name: 'tin_number')
  final String tinNumber;
  final int vat;
  @JsonKey(name: 'supplier_name')
  final String supplierName;
  @JsonKey(name: 'supplier_address')
  final String supplierAddress;
  @JsonKey(name: 'sub_category')
  final String subCategory;
  @JsonKey(name: 'invoice_number')
  final String invoiceNumber;

  Transaction({
    required this.id,
    required this.date,
    required this.type,
    required this.category,
    required this.description,
    required this.amount,
    this.paymentMethod = '',
    this.transactionNumber = '',
    this.receiptNumber = '',
    this.tinNumber = '',
    this.vat = 0,
    this.supplierName = '',
    this.supplierAddress = '',
    this.subCategory = '',
    this.invoiceNumber = '',
  });

  factory Transaction.fromJson(Map<String, dynamic> json) =>
      _$TransactionFromJson(json);

  Map<String, dynamic> toJson() => _$TransactionToJson(this);

  Transaction copyWith({
    int? id,
    String? date,
    TransactionType? type,
    String? category,
    String? description,
    double? amount,
    String? paymentMethod,
    String? transactionNumber,
    String? receiptNumber,
    String? tinNumber,
    int? vat,
    String? supplierName,
    String? supplierAddress,
    String? subCategory,
    String? invoiceNumber,
  }) {
    return Transaction(
      id: id ?? this.id,
      date: date ?? this.date,
      type: type ?? this.type,
      category: category ?? this.category,
      description: description ?? this.description,
      amount: amount ?? this.amount,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      transactionNumber: transactionNumber ?? this.transactionNumber,
      receiptNumber: receiptNumber ?? this.receiptNumber,
      tinNumber: tinNumber ?? this.tinNumber,
      vat: vat ?? this.vat,
      supplierName: supplierName ?? this.supplierName,
      supplierAddress: supplierAddress ?? this.supplierAddress,
      subCategory: subCategory ?? this.subCategory,
      invoiceNumber: invoiceNumber ?? this.invoiceNumber,
    );
  }
}

enum TransactionType {
  @JsonValue('revenue')
  revenue,
  @JsonValue('transaction')
  transaction,
  
  // Backward compatibility aliases
  @JsonValue('income')
  income,
  @JsonValue('expense')
  expense,
}

/// Revenue categories matching Next.js REVENUE_CATEGORIES
class RevenueCategories {
  static const String cash = 'Cash';
  static const String gcash = 'GCash';
  static const String grab = 'Grab';
  static const String paymaya = 'Maya';
  static const String creditCard = 'Credit Card';
  static const String others = 'Others';

  static const List<String> all = [cash, gcash, grab, paymaya, creditCard, others];
}

/// Transaction categories matching Next.js TRANSACTION_CATEGORIES
class TransactionCategories {
  static const String supplies = 'Supplies';
  static const String pastries = 'Pastries';
  static const String rent = 'Rent';
  static const String utilities = 'Utilities';
  static const String manpower = 'Manpower';
  static const String marketing = 'Marketing';
  static const String transpoAndDeliveries = 'Transpo and Deliveries';
  static const String rm = 'R&M';
  static const String mandatories = 'MANDATORIES';
  static const String pestControl = 'Pest control';
  static const String misc = 'MISC';
  static const String commissions = 'COMMISIONS';
  static const String taxes = 'TAXES';
  static const String others = 'OTHERS';

  static const List<String> all = [
    supplies,
    pastries,
    rent,
    utilities,
    manpower,
    marketing,
    transpoAndDeliveries,
    rm,
    mandatories,
    pestControl,
    misc,
    commissions,
    taxes,
    others,
  ];
}

// Type aliases for backward compatibility
typedef IncomeCategories = RevenueCategories;
typedef ExpenseCategories = TransactionCategories;

/// Payment methods matching Next.js PAYMENT_METHODS
class PaymentMethods {
  static const String cash = 'Cash';
  static const String check = 'Check';
  static const String bankTransfer = 'Bank Transfer';
  static const String creditCard = 'Credit Card';
  static const String gcash = 'GCash';
  static const String paymaya = 'Maya';
  static const String others = 'Others';

  static const List<String> all = [
    cash,
    check,
    bankTransfer,
    creditCard,
    gcash,
    paymaya,
    others,
  ];
}
